### ISSUE TITLE
prefer-const doesn't handle deconstructing right

### ISSUE THREAD

**Issue by KingDGrizzle** [eslint/eslint#10401]
Hello.

I'd like to request either a change in the `prefer-const` rule to properly handle deconstructing, or have an option object where we can disable it, just like `prefer-destructuring`.

**Tell us about your environment**

* **ESLint Version:** v4.19.1
* **Node Version:** 10.2.1
* **npm Version:** 6.1.0

**What parser (default, Babel-ESLint, etc.) are you using?**

Default

**Please show your full configuration:**

<details>
<summary>Configuration</summary>

```js
module.exports = {
	extends: "eslint:recommended",
	env: {
		node: true,
		es6: true,
	},
	parserOptions: {
		ecmaVersion: 9,
		sourceType: "module",
		ecmaFeatures: {
			experimentalObjectRestSpread: true,
		},
	},
	globals: {
		winston: true,
		Servers: true,
		Users: true,
		Gallery: true,
		Blog: true,
		Wiki: true,
		Database: true,
		Raw: true,
		configJS: true,
		configJSON: true,
	},
	plugins: [
		"promise",
		"node",
	],
	rules: {
		"no-extra-parens": ["warn", "all", { nestedBinaryExpressions: false	}],
		"accessor-pairs": "warn",
		"array-callback-return": "error",
		"curly": ["error", "multi-line", "consistent"],
		"dot-location": ["error", "property"],
		"dot-notation": "error",
		"eqeqeq": "error",
		"no-console": "error",
		"no-empty-function": "error",
		"no-floating-decimal": "error",
		"no-implied-eval": "error",
		"no-invalid-this": "error",
		"no-lone-blocks": "error",
		"no-multi-spaces": "error",
		"no-new-func": "error",
		"no-new-wrappers": "error",
		"no-new": "error",
		"no-octal-escape": "error",
		"no-return-assign": "error",
		"no-return-await": "error",
		"no-self-compare": "error",
		"no-sequences": "error",
		"no-unmodified-loop-condition": "error",
		"no-unused-expressions": ["error", { "allowShortCircuit": true }],
		"no-useless-call": "error",
		"no-useless-concat": "error",
		"no-useless-escape": "error",
		"no-void": "error",
		"no-warning-comments": "warn",
		"wrap-iife": "error",
		"yoda": "error",
		"no-label-var": "error",
		"no-shadow": "error",
		"no-undef-init": "error",
		"callback-return": "error",
		"handle-callback-err": "error",
		"no-mixed-requires": "error",
		"no-new-require": "error",
		"no-path-concat": "error",
		"array-bracket-spacing": "error",
		"block-spacing": "error",
		"brace-style": ["error", "1tbs", { allowSingleLine: true }],
		"comma-dangle": ["error", "always-multiline"],
		"comma-spacing": "error",
		"comma-style": "error",
		"computed-property-spacing": "error",
		"consistent-this": "error",
		"eol-last": "error",
		"func-names": "error",
		"func-style": ["error", "declaration", { allowArrowFunctions: true }],
		indent: ["error", "tab", { SwitchCase: 1 }],
		"key-spacing": "error",
		"keyword-spacing": "error",
		"max-depth": ["error", 8],
		"max-len": ["error", 200, 2],
		"max-nested-callbacks": ["error", { max: 4 }],
		"max-statements-per-line": ["error", { max: 2 }],
		"new-cap": "off",
		"newline-per-chained-call": ["error", { ignoreChainWithDepth: 3 }],
		"no-array-constructor": "error",
		"no-bitwise": "warn",
		"no-inline-comments": "error",
		"no-lonely-if": "error",
		"no-mixed-operators": "error",
		"no-multiple-empty-lines": ["error", { max: 2, maxEOF: 1, maxBOF: 0 }],
		"no-new-object": "error",
		"no-spaced-func": "error",
		"no-trailing-spaces": "error",
		"no-unneeded-ternary": "error",
		"no-whitespace-before-property": "error",
		"object-curly-newline": "error",
		"object-curly-spacing": ["error", "always"],
		"operator-assignment": "error",
		"operator-linebreak": ["error", "after"],
		"padded-blocks": ["error", "never"],
		"quote-props": ["error", "as-needed"],
		"quotes": ["error", "double", { avoidEscape: true, allowTemplateLiterals: true }],
		"semi-spacing": "error",
		"semi": "error",
		"space-before-blocks": "error",
		"space-before-function-paren": ["error", "always"],
		"space-in-parens": "error",
		"space-infix-ops": "error",
		"space-unary-ops": "error",
		"spaced-comment": "error",
		"unicode-bom": "error",
		"arrow-body-style": "error",
		"arrow-spacing": "error",
		"no-duplicate-imports": "error",
		"no-useless-computed-key": "error",
		"no-useless-constructor": "error",
		"prefer-arrow-callback": "error",
		"prefer-const": "error",
		"prefer-destructuring": ["error", {
			VariableDeclarator: {
				array: false,
				object: true,
			},
			AssignmentExpression: {
				array: true,
				object: true,
			},
		}, {
			enforceForRenamedProperties: false,
		}],
		"prefer-rest-params": "error",
		"prefer-spread": "error",
		"prefer-template": "error",
		"rest-spread-spacing": "error",
		"template-curly-spacing": "error",
		"yield-star-spacing": "error",
		"promise/no-return-wrap": "error",
		"promise/catch-or-return": "error",
		"promise/no-native": "off",
		"promise/no-nesting": "warn",
		"promise/no-promise-in-callback": "warn",
		"promise/no-callback-in-promise": "warn",
		"promise/no-return-in-finally": "warn",
		"node/no-missing-require": "off",
		"node/no-unpublished-bin": "error",
		"node/no-deprecated-api": "error",
		"node/exports-style": ["error", "module.exports"],
		"no-var": "error",
	},
};
```

</details>

**What did you do? Please include the actual source code causing the issue, as well as the command that you used to run ESLint.**

```js
let [variable1, ...numbers] = [1, 2, 3, 4, 5, 6];
numbers = numbers.reduce((a, number) => a + number, 0);
```

```bash
eslint ./file.js
```

**What did you expect to happen?**
prefer-const would not complain that variable1 isn't overwritten and would detect that it is deconstructed from an array.

**What actually happened? Please include the actual, raw output from ESLint.**
It complains that variable1 should use const, but then I cannot overwrite numbers due to it being const.
```bash
PS C:\Users\Vlad\Desktop\Development\> eslint ./file.js

C:\Users\Vlad\Desktop\Development\file.js
  56:8  error  'variable1' is never reassigned. Use 'const' instead  prefer-const

✖ 1 problem (1 error, 0 warnings)
```

**Comment by platinumazure** [comment]
Hi @KingDGrizzle, thanks for the issue.

I think this is working as designed, but the reason might be a little surprising. In array destructuring, you actually can have "empty slots" in the destructure array for values you don't need. So you could actually do this:

```js
let [, ...numbers] = [1, 2, 3, 4, 5, 6];
numbers = numbers.reduce((a, number) => a + number, 0);
```

For cases where you do need `variable1` elsewhere, you may wish to use the [`destructuring: "all"` option](https://eslint.org/docs/rules/prefer-const#destructuring). In that case, the code should detect that `numbers` is mutated, so `variable1` will be ignored.

Better still, though, would be to create a new variable name for `numbers` (e.g., `sum`), and have everything be `const`. But you may have good reasons for reusing the variable name :smile:

**Comment by KingDGrizzle** [comment]
Hey.

I know about deconstructing arrays and having holes (god send when executing a regex on a string), and the code given was, in the end, an example code.

I must of missed the options part of the rule when going to the webpage and seeing what it does. Woops.

Indeed, changing it to "all" fixes this, why it isn't defaulting to all is another question for another time 😄 

Thanks for your time! I'll leave this open as it's still a suggestion to change the default (which would make more sense, as this also affects things like 
```js
for (let [key, value] of Object.entries(someObject)) {
  key = key.toLowerCase();
  // ...do something with key - value
}
```
).

**Comment by platinumazure** [comment]
In most cases, you could do multiple assignments to cover this:

```js
const foo = arr[0];
let [, bar] = arr;
// Reassign bar later
```

Or you could create a new variable:

```js
for (const [key, value] of Object.entries(someObject)) {
  const lcKey = key.toLowerCase();
  // ...do something with lcKey - value
}
```

And if you are normally okay with the default behavior but want to allow one-off exceptions for loops, etc., you could use disable comments:

```js
// eslint-disable-next-line prefer-const
for (let [key, value] of Object.entries(someObject)) {
  key = key.toLowerCase();
  // ...do something with lcKey - value
}
```

**Comment by KingDGrizzle** [comment]
I guess that's true.

I'll leave this up to you then if this issue can be closed or not, as my "problem" was solved. 😃

**Comment by kaicataldo** [comment]
Closing this issue as it looks like the question has been answered. Please feel free to visit us in the [ESLint Gitter](https://gitter.im/eslint/eslint) if you have any other issues!
